### PR TITLE
Add a section on record/replay tests to design.md

### PR DESCRIPTION
**Pull request by vangent**
#276

### CODE REVIEW

**Comment by vangent** [comment]
Fixes #276

File: internal/docs/design.md [hunk 1]
```diff
@@ -104,33 +99,113 @@ https://godoc.org/github.com/google/go-cloud/blob#Bucket.NewWriter
...
+### Record Mode
+
+See https://github.com/google/go-cloud/issues/128 for active work on improving
+this.
+
+1.  Provision resources.
+
+    -   For example, the tests for the AWS implementation of Blob requires a
+        bucket to be provisioned.
+    -   TODO: Use Terraform scripts to provision the resources needed for a
```
Comment (shantuo) [inline internal/docs/design.md]: Add the issue number here, "TODO(issue #)"

File: internal/docs/design.md [hunk 2]
```diff
@@ -104,33 +99,113 @@ https://godoc.org/github.com/google/go-cloud/blob#Bucket.NewWriter
...
+
+    -   TODO: The test will read the Terraform output file `output.tf` to find
+        its inputs.
+    -   For now, pass the required resources via test-specific flags. In some
+        cases, tests are
+        [hardcoded to specific resource names](https://github.com/google/go-cloud/issues/286).
+
+3.  The test will save the network interactions for subsequent replays.
+
+    -   TODO: The test will save the `output.tf` file in order to replay using
```
Comment (shantuo) [inline internal/docs/design.md]: Same, add the issue number.

To be clear, output.tf contains the variable definitions terraform outputs. One needs to run `terraform output` with a `.tfstate` file (not checked in, and associated with the `terraform apply` run) to get the actual output. What we want to save here is the actual output, not `output.tf`.

File: internal/docs/design.md [hunk 3]
```diff
@@ -104,33 +99,113 @@ https://godoc.org/github.com/google/go-cloud/blob#Bucket.NewWriter
...
+    malformed request, bad permissions) where appropriate so that application
+    logic does not attempt to retry idempotent operations, but the
+    responsibility is largely on the library, not on the application.
 
 [cascading failure]:
 https://landing.google.com/sre/book/chapters/addressing-cascading-failures.html
+
+## Tests
+
+For portable APIs, the distinction between developer and operator is blurred for
```
Comment (zombiezen) [inline internal/docs/design.md]: Don't mention the roles here. The roles are focused on the user of the library, not the maintainer of the library.

File: internal/docs/design.md (hunk 1 above)
Comment (vangent) [reply internal/docs/design.md]: Done.

File: internal/docs/design.md (hunk 2 above)
Comment (vangent) [reply internal/docs/design.md]: Done, and updated to remove reference to output.tf explicitly.

File: internal/docs/design.md (hunk 3 above)
Comment (vangent) [reply internal/docs/design.md]: Done.

**Pull request merged**